src/core/mesh/qgstriangularmesh.cpp

```diff
@@ -165,3 +177,31 @@ const QVector<int> &QgsTriangularMesh::trianglesToNativeFaces() const
   return mTrianglesToNativeFaces;
 }
 
+int QgsTriangularMesh::faceIndexForPoint( const QgsPointXY &point ) const
+{
+  const QList<QgsFeatureId> face_indexes = mSpatialIndex.intersects( QgsRectangle( point, point ) );
+  for ( QgsFeatureId fid : face_indexes )
+  {
+    int face_index = static_cast<int>( fid );
+    const QgsMeshFace &face = mTriangularMesh.faces.at( face_index );
+    const QgsGeometry geom = QgsMeshUtils::toGeometry( face, mTriangularMesh.vertices );
+    if ( geom.contains( &point ) )
+      return face_index;
+  }
+  return -1;
+}
+
+QgsGeometry QgsMeshUtils::toGeometry( const QgsMeshFace &face, const QVector<QgsMeshVertex> &vertices )
+{
+  QVector<QgsPointXY> ring;
```

Comment:
Would be more efficient to avoid the QgsPointXY methods here. Instead I'd suggest

    QVector<QgsPoint> ring;
    ...
    
    std::unique_ptr< QgsPolygon > polygon = qgis::make_unique< QgsPolygon >();
    polygon.setExteriorRing( new QgsLineString( ring ) );
    return QgsGeometry( std::move( polygon ) );